### PR TITLE
Add traceability endpoint

### PR DESCRIPTION
**Pull request by eugenioLopezRamos**
- Adds a traceability endpoint
- Adds tests

### CODE REVIEW

File: app/controllers/traceability_controller.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+class TraceabilityController < ApplicationController
+  skip_before_action :verify_authenticity_token
+  before_action :verify_token, only: :endpoints_info
+
+  def endpoints_info
+    render json: { services: Service.all.map(&method(:payload_for_service)) }
+  end
+
+  def get_token
```
Comment (leosoto) [inline app/controllers/traceability_controller.rb]: I don't understand why we need this

File: test/controllers/traceability_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
+require 'test_helper'
+
+class StaticPagesControllerTest < ActionDispatch::IntegrationTest
```
Comment (leosoto) [inline test/controllers/traceability_controller_test.rb]: StaticPages?

File: app/controllers/traceability_controller.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    return hsh if service.public?
+    hsh[:token] = service.generate_client_token
+    return hsh unless hsh[:token].nil?
+    service.generate_provider_credentials
+    hsh[:token] = service.generate_client_token
+    hsh
+  end
+
+  private
+  def verify_token
```
Comment (leosoto) [inline app/controllers/traceability_controller.rb]: Why do we need this too? Wouldn't it be easier to write the ` return head(401) if params[:secret] != ENV['TRAZABILIDAD_SECRET']` here and be done with it (without issuing nor validating useless JWT tokens)

File: config/routes.rb [hunk 4]
```diff
@@ -7,6 +7,9 @@
   devise_for :users, :controllers => { :omniauth_callbacks => "users/omniauth_callbacks" }
   root to: 'home#root'
 
+  post 'trazabilidad/token', to: 'traceability#get_token'
```
Comment (leosoto) [inline config/routes.rb]: Again, don't see the need for this endpoint

File: testft [hunk 5]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+voltos run docker-compose run web rails test test/features/$1_test.rb
```
Comment (leosoto) [inline testft]: If you add this script it would be better to do it on a separate PR alongside — at least — a note on the README.

File: app/controllers/traceability_controller.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    render json: { token: SignerApi.encode_token(payload,
+                                                 ENV['TRAZABILIDAD_SECRET']) }
+  end
+
+  def payload_for_service(service)
+    hsh = { url: service&.service_versions&.current&.first&.base_url }
+    return hsh if service.public?
+    hsh[:token] = service.generate_client_token
+    return hsh unless hsh[:token].nil?
+    service.generate_provider_credentials
```
Comment (leosoto) [inline app/controllers/traceability_controller.rb]: Wrong solution. This method will not persist (nor communicate) the provider credentials to anyone.

What problem are you trying to solve? Did you find non-public services without provider credentials? How could that happen?

File: app/controllers/traceability_controller.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    return head(401) if params[:secret] != ENV['TRAZABILIDAD_SECRET']
+    payload = { exp: (Time.now + 1.day).to_i, name: params[:user] }
+    render json: { token: SignerApi.encode_token(payload,
+                                                 ENV['TRAZABILIDAD_SECRET']) }
+  end
+
+  def payload_for_service(service)
+    hsh = { url: service&.service_versions&.current&.first&.base_url }
+    return hsh if service.public?
+    hsh[:token] = service.generate_client_token
```
Comment (leosoto) [inline app/controllers/traceability_controller.rb]: This will work, but the generated JWT token will have the wrong `sub`ject. The `Service#generate_client_token` is meant to generate tokens to be used on the console, usually by the service provider himself. That's why the `sub` in `Service#generate_client_token` is populated with the id of the organization *providing* the service. 

But you are generating tokens for a _different_  subject: The "trazabilidad" system. In order for the service to properly identify who is getting access to its system, we should put a different `sub` on the JWT. Which one? No idea. Maybe the URL of the trazabilidad system?

In any case, you will have to pass in a different `sub`ject to `generate_client_token` (which still has to default to the organization url for when it is used from the console)

File: app/controllers/traceability_controller.rb (hunk 6 above)
Comment (eugenioLopezRamos) [reply app/controllers/traceability_controller.rb]: It was explained to me that there could be non public services where credentials where not generated, and that I'd need to create them using that method before being able to generate_client_token

Comment (leosoto) [reply app/controllers/traceability_controller.rb]: I still wonder how that could happen. If you have an answer to that, we can try to work on how to solve it.

Because, as I said before, the method you are calling is not solving the (eventual) problem.
